### ISSUE TITLE
text/plain mime type when using c.JSON

### ISSUE THREAD

**Issue by kokaz** [gin-gonic/gin#593]
Hi,

I'm using master branch at commit 5caaac4, and when my `BindJSON()` fails because no data:

```
curl -i -X POST http://localhost:8180/v1/feeds
```

``` go
func PostFeed(c *gin.Context) {
    var in = &newFeed{}
    err := c.BindJSON(&in)
    if err != nil {
        c.JSON(-1, gin.H{
            "status":  http.StatusBadRequest,
            "error":   err.Error(),
            "message": "Your payload is not valid or well-formated.",
        })
        return
    }
}
```

The `Content-Type` returned is `text/plain; charset=utf-8` I logged the `Content-Type` before and after the call to `c.JSON` with `c.Writer.Header.Get()` it's OK but no in my final response. I don't have any middleware (nor the `gin.Default()` one) adding/editing my `Content-Type`. I looked through the whole source of code gin without finding anything!

Any idea?

**Comment by kokaz** [comment]
After hours of debugging I'm quite sure there is a bug in Gin that cause that. If I call `c.Header("HEADER", "VALUE")` before `c.BindJSON` (in the case where it fails) all the headers set are kept (good) **BUT** if I do the same in the `if` with `err != nil` all headers set before and after `c.JSON` are swept away without consent (bad)! I'm still investing where it could fail.

**Comment by kokaz** [comment]
Unless I did something bad I found (thanks git bisect) that the incriminated commit is 4c4444b16033c7b000492f57119901cba9870b5c. After reverting it and explicitly call `c.Writer.WriteHeaderNow()` everything is working.

So now after some thought I am a bit confused how the thing is designed because the only thing I want is use `c.Bind*` and print a JSON describing the error (as any RESTful API) and the fact the `Bind` funcs are themselves calling `c.Abort*` (with the `WriteHeaderNow` call) are pretty blocking for use case like that.
